### PR TITLE
Implement prune method

### PR DESCRIPTION
**Pull request by janekolszak**
- Update jest to v29 and add ts-jest
- Add prune method

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -26,3 +26,6 @@ yalc.lock
 bundles/
 
 .secrets
+
+./tools/src
+./tools/dst
```
Comment (ppedziwiatr) [inline .gitignore:31]: new line :-) https://stackoverflow.com/a/5813359

File: package.json [hunk 2]
```diff
@@ -26,6 +26,7 @@
     "postversion": "git push && git push --tags",
     "yalc:publish": "yarn build && yalc publish --push",
     "test": "jest",
+    "test:real": "jest ./src/__tests__/lmdb-real-data",
```
Comment (ppedziwiatr) [inline package.json:29]: again, I'm not sure if this test should be run by default (just like in the leveldb)

File: src/LmdbCache.ts [hunk 3]
```diff
@@ -100,4 +101,63 @@ export class LmdbCache<V = any> implements SortKeyCache<V> {
   storage<S>(): S {
     return this.db as S;
   }
+
+  async prune(entriesStored: number = 1): Promise<PruneStats> {
```
Comment (ppedziwiatr) [inline src/LmdbCache.ts:105]: `yarn lint` please :-)

File: src/LmdbCache.ts [hunk 4]
```diff
@@ -100,4 +101,63 @@ export class LmdbCache<V = any> implements SortKeyCache<V> {
...
+
+    return {
+      entriesBefore: statsBefore.entryCount,
+      sizeBefore: statsBefore.mapSize,
+      entriesAfter: statsAfter.entryCount,
+      sizeAfter: statsAfter.mapSize
+    } as PruneStats
+  }
+
+  async forEach(callback: (k: CacheKey, v: any) => Promise<void>) {
```
Comment (ppedziwiatr) [inline src/LmdbCache.ts:148]: hmm, where is it used? the prune method seems to use standard Array's `forEach` method...

File: src/LmdbCache.ts [hunk 5]
```diff
@@ -100,4 +101,63 @@ export class LmdbCache<V = any> implements SortKeyCache<V> {
...
+    await this.db.flushed
+
+    const statsAfter: any = this.db.getStats()
+
+    return {
+      entriesBefore: statsBefore.entryCount,
+      sizeBefore: statsBefore.mapSize,
+      entriesAfter: statsAfter.entryCount,
+      sizeAfter: statsAfter.mapSize
+    } as PruneStats
```
Comment (ppedziwiatr) [inline src/LmdbCache.ts:145]: why "as" is needed here?

File: src/LmdbCache.ts (hunk 3 above)
Comment (ppedziwiatr) [reply src/LmdbCache.ts:105]: I'm used to `;` at the end of line :-( but not sure if `yarn lint` will handle this...

File: src/__tests__/lmdb-cache-prune.test.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    const contracts = 1
+    const entriesPerContract = 10
+    const sut = await cache(contracts, entriesPerContract)
+    expect(await sut.prune(1)).toMatchObject({ entriesBefore: contracts * entriesPerContract, entriesAfter: contracts * 1 })
+  });
+
+  it('should remove all unneeded entries, in many contracts', async () => {
+    const contracts = 200
+    const entriesPerContract = 10
+    const sut = await cache(contracts, entriesPerContract)
```
Comment (ppedziwiatr) [inline src/__tests__/lmdb-cache-prune.test.ts:67]: same comment as in the lmdb - this test should verify whether the proper amount of **latest** keys was stored for each contract...

File: src/__tests__/lmdb-real-data.test.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { LmdbCache } from '../LmdbCache';
+import { defaultCacheOptions, PruneStats } from 'warp-contracts';
+
+describe.skip('Prune real Lmdb cache. Copy the real cache to ./cache, unskip and run npm run test:real', () => {
```
Comment (ppedziwiatr) [inline src/__tests__/lmdb-real-data.test.ts:4]: oh, this one have 'skip' - I believe the leveldb version was missing the skip :-)

File: tools/copycache.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
+/* eslint-disable */
+import { LevelDbCache, defaultCacheOptions } from 'warp-contracts';
+import { LmdbCache } from '../src/LmdbCache'
+
+async function main() {
+  console.log('Copy LMDB cache to LevelDB');
+  const src = new LmdbCache<any>({ ...defaultCacheOptions, dbLocation: "./src" });
+  const dst = new LevelDbCache<any>({ ...defaultCacheOptions, dbLocation: "./dst" })
```
Comment (ppedziwiatr) [inline tools/copycache.ts:8]: is ./dst in the gitignore?

File: src/__tests__/lmdb-cache-prune.test.ts [hunk 9]
```diff
@@ -0,0 +1,71 @@
+import { LmdbCache } from '../LmdbCache';
+import { defaultCacheOptions, PruneStats } from 'warp-contracts';
+import * as fs from 'fs';
+
+describe('Lmdb cache prune', () => {
+  beforeEach(() => {
+    fs.rmSync('./cache', { force: true, recursive: true });
```
Comment (ppedziwiatr) [inline src/__tests__/lmdb-cache-prune.test.ts:7]: are you sure that it will not fail if the ./cache dir is not created?

**Review by ppedziwiatr — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (janekolszak) [reply package.json:29]: It's already skipped

File: src/LmdbCache.ts (hunk 4 above)
Comment (janekolszak) [reply src/LmdbCache.ts:148]: It's used in the script for copying lmdb cache to leveldb

File: src/LmdbCache.ts (hunk 5 above)
Comment (janekolszak) [reply src/LmdbCache.ts:145]: not needed

File: src/__tests__/lmdb-cache-prune.test.ts (hunk 9 above)
Comment (janekolszak) [reply src/__tests__/lmdb-cache-prune.test.ts:7]: That's what `force` is for. On Unix if you ad force removing it skips errors

File: tools/copycache.ts (hunk 8 above)
Comment (janekolszak) [reply tools/copycache.ts:8]: added

File: src/__tests__/lmdb-real-data.test.ts (hunk 7 above)
Comment (janekolszak) [reply src/__tests__/lmdb-real-data.test.ts:4]: yeap, my bad

File: src/LmdbCache.ts (hunk 4 above)
Comment (ppedziwiatr) [reply src/LmdbCache.ts:150]: Can we move it then to the script then..? It looks kinda weird here...

**Review by ppedziwiatr — APPROVED** [review]

**Pull request merged**